### PR TITLE
numpy like nonzero (called nonzero_tuple)

### PR DESCRIPTION
**Pull request by umanwizard**
No performance degradation compared to Numpy when indexing:

```
In [15]: x=torch.randn((1000,1000))                                                                                                                                                         

In [16]: %timeit x[x.nonzero_tuple()]                                                                                                                                                       
4.63 ms ± 102 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [17]: y=x.numpy()                                                                                                                                                                        

In [18]: %timeit y[y.nonzero()]                                                                                                                                                             
14.6 ms ± 281 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [20]: x=x.t()                                                                                                                                                                            

In [22]: %timeit x[x.nonzero_tuple()]                                                                                                                                                       
9.01 ms ± 626 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [24]: y=x.numpy()                                                                                                                                                                        

In [25]: %timeit y[y.nonzero()]                                                                                                                                                             
16.8 ms ± 770 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

```

### CODE REVIEW

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -3491,6 +3491,32 @@ def merge_dicts(*dicts):
             [ 3,  3]])
 """)
 
+add_docstr(torch.nonzero_tuple,
```
Comment (SsnL) [inline torch/_torch_docs.py]: I wonder if it is better to add a kwarg to `nonzero`.

Comment (umanwizard) [reply torch/_torch_docs.py]: @SsnL I don't think it makes sense to have a kwarg controlling the output type (tuple vs. tensor)

Comment (SsnL) [reply torch/_torch_docs.py]: Hmm, but we have several existing examples: torch.lu, torch.unique, torch.unique_consecutive

**Comment by gchanan** [comment]
that benchmark seems good, but wouldn't I also want to know the straight up runtime of the old-vs-the-new; you don't _always_ index after a nonzero.

**Comment by umanwizard** [comment]
@gchanan it should be essentially the same since unbind is free (modulo overhead)

```
In [4]: %timeit x.nonzero()                                                                                                                                                                                                                   
3.72 ms ± 113 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

In [6]: %timeit x.nonzero_tuple()                                                                                                                                                                                                             
3.67 ms ± 87.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
```

File: test/test_torch.py [hunk 2]
```diff
@@ -8828,6 +8828,14 @@ def gen_nontrivial_input(num_src, dtype, device):
                             self.assertNotEqual(tensor[dst1[i, 0], dst1[i, 1], dst1[i, 2]].item(), 0)
                         lex = is_lexicographically_sorted(dst1)
                         self.assertEqual(torch.ones_like(lex), lex)
+                    if TEST_NUMPY:
```
Comment (colesbury) [inline test/test_torch.py]: This needs a test for the empty tensor (can be added to `test_nonzero_empty` below). It also needs a test for a zero-dim (scalar) tensor (i.e. `torch.tensor(1)` and `torch.tensor(0)`).

**Review by colesbury — COMMENTED** [review]
To echo @SsnL's point: a kwarg is consistent with what we've done in the past, discussed in https://github.com/pytorch/pytorch/issues/2739, and with NumPy's [unique](https://docs.scipy.org/doc/numpy-1.16.1/reference/generated/numpy.unique.html#numpy.unique) function.

File: torch/_torch_docs.py (hunk 1 above)
Comment (gchanan) [reply torch/_torch_docs.py]: I agree with @SsnL here -- otherwise we are going to have non-numpy behavior forever for no good reason.

How about something like this:
- add a new parameter (return_tuple?) defaulting to None.
- None for now means old behavior, you can turn on a deprecation warning on None, saying None will change to mean True in the next release, if they want old behavior, pass False, otherwise pass True.
- In the next release, change the default.

Comment (vadimkantorov) [reply torch/_torch_docs.py]: If `nonzero` default behavior is okayed to change, #1834 is also related. NumPy returns a squeezed tensor (differently from PyTorch): 
```python
np.random.rand(10).nonzero()[0].shape
# (10,)
```

File: test/test_torch.py (hunk 2 above)
Comment (gchanan) [reply test/test_torch.py]: Case that seems to be different than numpy:
```
>>> torch.nonzero_tuple(torch.tensor(0))
()

>>> torch.nonzero_tuple(torch.tensor(5))
()

>>> np.array(5).nonzero()
(array([0]),)

>>> np.array(0).nonzero()
(array([], dtype=int64),)
```

File: torch/_torch_docs.py (hunk 1 above)
Comment (fmassa) [reply torch/_torch_docs.py]: @vadimkantorov this PR does exactly that, and follows numpy semantics

File: test/test_torch.py (hunk 2 above)
Comment (umanwizard) [reply test/test_torch.py]: I think this is a bug in numpy. It doesn't match their documentation: "Returns a tuple of arrays, **one for each dimension of a**" ... I think it's semantically correct to do it the way I have it now, but I can see the argument for bug-for-bug compatibility with np... what do you guys think?

File: torch/_torch_docs.py (hunk 1 above)
Comment (umanwizard) [reply torch/_torch_docs.py]: It seems there is a consensus here, so I will make this a kwarg.

Comment (umanwizard) [reply torch/_torch_docs.py]: It seems the way we are supporting multiple possible return types (torch.lu , torch.unique, etc.) is by returning a tuple of (tensor, tensor, tensor) in C++, then extracting values as necessary in Python:

```python
def unique_consecutive(input, return_inverse=False, return_counts=False, dim=None):
    # snip
    output, inverse_indices, counts = torch._C._VariableFunctions.unique_consecutive(
        input, return_inverse=return_inverse, return_counts=return_counts, dim=dim)
    if return_inverse and return_counts:
        return output, inverse_indices, counts
    if return_inverse:
        return output, inverse_indices
    if return_counts:
        return output, counts
    return output
```

So, the logical way to implement this would be to make `nonzero` return a `std::tuple<Tensor, TensorList>` in C++ (it currently just returns `Tensor`), and then extract the correct thing in python. But that would be a breaking change...

Perhaps we could keep both `nonzero` and `nonzero_tuple` in C++, and have them merged into one thing (with a kwarg) in python?

Comment (gchanan) [reply torch/_torch_docs.py]: > Perhaps we could keep both nonzero and nonzero_tuple in C++, and have them merged into one thing (with a kwarg) in python?

 That sounds good to me.

File: test/test_torch.py (hunk 2 above)
Comment (fmassa) [reply test/test_torch.py]: I think we should keep numpy behavior.
When performing advanced indexing with a scalar boolean tensor containing True, it should return the same as if we passed bool True, which returns all the dimensions.

Comment (gchanan) [reply test/test_torch.py]: @fmassa I'm not sure how to interpret your comment.

How is boolean indexing relevant here?

Doesn't `True` insert a dimension?

Comment (fmassa) [reply test/test_torch.py]: @gchanan scratch what I just said. Upon checking [the numpy detailed notes of advanced indexing](https://www.numpy.org/devdocs/reference/arrays.indexing.html#detailed-notes)
> the nonzero equivalence for Boolean arrays does not hold for zero dimensional boolean arrays

So what I said doesn't hold.

---
For completeness, I was trying to make a point that the following should hold:
```python
x = a[np.array(True)]
y = a[True]
z = a[np.array(True).nonzero()]
# this is true
x == y
# but z is handled differently by numpy
x != z
```

File: torch/functional.py [hunk 3]
```diff
@@ -934,6 +935,65 @@ def btrisolve(b, LU_data, LU_pivots, out=None):
                   "removed in the next release. Please use torch.lu_solve instead.", stacklevel=2)
     return torch.lu_solve(b, LU_data=LU_data, LU_pivots=LU_pivots, out=out)
 
+def nonzero(input, **kwargs):
+    r"""
+    nonzero(input, *, out=None, as_tuple=False) -> LongTensor or tuple of LongTensors
```
Comment (gchanan) [inline torch/functional.py]: What do you think about doing all the parsing at the python arg parser level?  That way we are guaranteed to treat everything consistently.

**Review by gchanan — COMMENTED** [review]
you seem to have some third_party changes, that shouldn't be in here.

File: torch/tensor.py [hunk 4]
```diff
@@ -324,6 +324,10 @@ def lu(self, pivot=True, get_infos=False):
         else:
             return LU, pivots
 
+    def nonzero(self, as_tuple=False):
```
Comment (gchanan) [inline torch/tensor.py]: doesn't this mean we accept as_tuple as a non-kwarg?

File: torch/functional.py (hunk 3 above)
Comment (umanwizard) [reply torch/functional.py]: Right now there is no support there afaik for returning different result types depending on the value of a kwarg, so it would have to be significantly refactored.

File: torch/tensor.py (hunk 4 above)
Comment (umanwizard) [reply torch/tensor.py]: fixed

File: torch/functional.py (hunk 3 above)
Comment (gchanan) [reply torch/functional.py]: can't you just handwrite the parsing code?  At the end of the day, the parsing related code just returns a PyObject, so this should be fine, no?

File: aten/src/ATen/native/Indexing.cpp [hunk 5]
```diff
@@ -569,4 +569,20 @@ Tensor _gather_sparse_backward(const Tensor& self, int64_t dim, const Tensor& in
...
 
+std::vector<Tensor> nonzero_tuple(const Tensor& self) {
+  // special case scalar for compatibility with numpy:
+  //
+  // >>> np.array(5).nonzero()
+  // (array([0]),)
+  // >>> np.array(0).nonzero()
+  // (array([], dtype=int64),)
+  if (self.dim() == 0) {
+    if (at::is_nonzero(self)) {
```
Comment (gchanan) [inline aten/src/ATen/native/Indexing.cpp]: I think you can just unsqueeze in this case to get the correct behavior.

File: aten/src/ATen/native/native_functions.yaml [hunk 6]
```diff
@@ -3096,6 +3096,9 @@
 - func: nonzero(Tensor self) -> Tensor
   variants: method, function
 
+- func: nonzero_tuple(Tensor self) -> Tensor[]
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: it's a little strange to call this nonzero_tuple when we don't return a tuple (in C++), and the behavior isn't just "return a tuple" (because of numpy compatbility concerns.  Maybe just call this nonzero_numpy?

File: aten/src/ATen/native/native_functions.yaml [hunk 7]
```diff
@@ -3096,6 +3096,9 @@
...
+  variants: method, function
```
Comment (gchanan) [inline aten/src/ATen/native/native_functions.yaml]: one strange thing here related to the parsing is that this nonzero has an out variant, but nonzero_tuple doesn't.  That seems fine because tuples are immutable in python anyway, but it would be nice to at least test passing an `out` from python and to see that we get a reasonable error message.

File: tools/pyi/gen_pyi.py [hunk 8]
```diff
@@ -63,6 +63,7 @@
     'margin_ranking_loss',
     'triplet_margin_loss',
     # Somehow, these are defined in both _C and in functional. Ick!
+    'nonzero',
```
Comment (gchanan) [inline tools/pyi/gen_pyi.py]: can you explain this?  nonzero is defined in both _C and functional?  why?

File: torch/_tensor_docs.py [hunk 9]
```diff
@@ -1658,11 +1658,11 @@ def callable(a, b) -> number
 Alias for :meth:`~Tensor.numel`
 """)
 
-add_docstr_all('nonzero',
+add_docstr_all('nonzero_tuple',
```
Comment (gchanan) [inline torch/_tensor_docs.py]: I don't think we should expose this to python.

File: torch/_torch_docs.py [hunk 10]
```diff
@@ -3458,38 +3458,12 @@ def merge_dicts(*dicts):
     tensor([-0.0090,  0.2262,  0.0682,  0.2866, -0.3940])
 """)
 
-add_docstr(torch.nonzero,
+add_docstr(torch.nonzero_tuple,
            r"""
-nonzero(input, out=None) -> LongTensor
+.. function:: nonzero_tuple(self) -> tuple of Tensors
```
Comment (gchanan) [inline torch/_torch_docs.py]: same here, I don't think we should expose this to python.

File: torch/functional.py [hunk 11]
```diff
@@ -20,6 +20,7 @@
     'lu_unpack',
     'norm',
     'meshgrid',
+    'nonzero',
```
Comment (gchanan) [inline torch/functional.py]: why is this in functional?

File: torch/tensor.py [hunk 12]
```diff
@@ -324,6 +324,10 @@ def lu(self, pivot=True, get_infos=False):
         else:
             return LU, pivots
 
+    def nonzero(self, **kwargs):
```
Comment (gchanan) [inline torch/tensor.py]: I think if you moved all the parsing to C++ you could get rid of this.

**Review by gchanan — CHANGES_REQUESTED** [review]

File: aten/src/ATen/native/Indexing.cpp (hunk 5 above)
Comment (umanwizard) [reply aten/src/ATen/native/Indexing.cpp]: Is this what you mean? It doesn't give the correct behavior.

```python
>>> torch.tensor(5).unsqueeze(0)
tensor([5])
>>> np.array(5).nonzero()
(array([0]),)
```

Comment (gchanan) [reply aten/src/ATen/native/Indexing.cpp]: I meant `unsqueeze` before you call into nonzero.

File: torch/_torch_docs.py [hunk 13]
```diff
@@ -3471,10 +3473,26 @@ def merge_dicts(*dicts):
 :attr:`out` is of size :math:`(z \times n)`, where :math:`z` is the total number of
 non-zero elements in the :attr:`input` tensor.
 
+**When** :attr:`as_tuple` **is true:**
+
+Returns a tuple of 1-D tensors, one for each dimension in :attr:`input`,
+each containing the indices (in that dimension) of all non-zero elements of
+:attr:`input` .
+
+If :attr:`input` has `n` dimensions, then the resulting tuple contains `n` tensors
```
Comment (gchanan) [inline torch/_torch_docs.py]: shouldn't this explain the 0-dimensional exception case?

Comment (umanwizard) [reply torch/_torch_docs.py]: Yes, good point

**Review by gchanan — APPROVED** [review]
There's a couple of unresolved comments here, but overall looks good.

File: aten/src/ATen/native/Indexing.cpp (hunk 5 above)
Comment (umanwizard) [reply aten/src/ATen/native/Indexing.cpp]: (y)
